### PR TITLE
[cosi] Make each node verify the challenge 

### PR DESCRIPTION
**Pull request by nikkolasg**
This PR brings an additional important check for the cosigners. Each cosigners should perform a verification on the challenge received by the leader, otherwise, there is no guarantee about what they are signing. This PR brings the following changes:
+ the statement S to sign is now always included in the announcement. This requirement may be relaxed later but for the moment, this is the only way to make the draft consistent. (we can't send the  statement during the challenge phase, otherwise each node that refuses to sign will make the protocol stop).
+ each node performs a syntactic and semantic check on the statement S during the commitment phase. If the check output a negative result, then the node quit the protocol.
+ the leader sends the challenge AND the aggregate commitment during the challenge phase
+ each node performs a check on the received challenge, namely they reconstruct `c' = H(R || A || S)` and check `c' == c`.

### CODE REVIEW

**Comment by bford** [comment]
Good, thanks Nicolas.

One consistency issue with the writing:  The draft currently uses the letter 'M' to represent a set of participants, which is confusing and inconsistent with common usage where 'M' almost always represents the *Message* to be signed.  Note that RFC 8032 uses M to mean Message, as per this common practice, and we should definitely remain consistent with that since that's clearly the most closely-related document.

It looks like you're using both N and M to represent (slightly-different) participant sets?  If you need two participant sets, why not call one N and the other N' for example?

And N isn't a great letter for a participant set either; usually N represents a number (perhaps a number of participants, but a *number* not a *set*).  Is the letter P used for something else at the moment?  Perhaps P and P' would be a better letter to use for participant sets?

**Comment by Daeinar** [comment]
The letter P seems not to be used at the moment. So we can use P and P' for the set of participants and then switch to M as the message / statement. Moreover, I noticed that we switch back and forth between 'nodes' and 'participants' to denote the co-signers. We should probably be consistent there as well. Any preferences?

**Comment by Daeinar** [comment]
After briefly talking to @nikkolasg, we use 'participants' for flat CoSi and 'nodes' for the tree-version.

**Comment by Daeinar** [comment]
To avoid misunderstandings between the messages (announcement, commitment, challenge, response) that are sent during the CoSi protocol and the message M that gets signed, should we go with 'the statement M' instead of 'the message M'?

**Comment by bford** [comment]
Yes, 'statement M' is a bit awkward but perhaps a reasonable compromise.  Maybe say 'statement message M' at least the first time it's used, and then just 'statement M' thereafter?

It might also be good to clarify the terms used for the announce/commit/etc messages as well: e.g., perhaps call them 'CoSi messages' or (or even 'CoSi protocol messages' the first time) instead of just 'messages'?

**Comment by nikkolasg** [comment]
What about keeping `message M` but finding another name for the announcement / commitments etc ? I'm not a fan of `statement M`, I find it removes one ambiguity to introduce another.. (why not go with `statement S` then ;) 
We could for example describe these four CoSi protocol messages as "packets" in the beginning and then refer to them later as `the Announcement packet` etc.
```
The CoSi protocol makes use of four distinct packets, one for each phase: 
+ the Announcement packet,
+ the Commitment packet, 
+ the Challenge packet
+ the Response packet.
```
Otherwise looking at other RFCs, people often use the word "records" to describe actual "structured packets" that goes into the network. Even though I agree that feels a bit unnatural, it may be worth considering.

**Comment by Daeinar** [comment]
I'm fine with `message M` and `CoSi packet` or `CoSi record`.

@nikkolasg: Why do you think records are unnatural? TLS for example also uses this notion: https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_record

**Comment by bford** [comment]
The alternative term 'packet' had also occurred to me as well, but I didn't suggest it because it sounds a bit weird in this context - but at the same time it would be consistent with many other uses of the term 'packet' (including in the OpenPGP protocol etc), so given that both of thought of it independently maybe that's not a bad idea.

'record' is also fine with me in principle, but I don't know if it's any better than 'packet'.

**Comment by nikkolasg** [comment]
I picked the word records exactly from the TLS 1.3 RFC but tp me it sounds more like a record is a structured list of constant sized fields. Our RFC uses protobuf to marshal our packet, so to me, a protobuf-marshalled packet (the actual slice of bytes) is a record, but the protobuf definition is not. Packet sounds more general.
In the end, it's only nitpicking, I'm ok with both ;)

**Comment by bford** [comment]
Agreed, I think 'packet' is more generally and marginally better in this context; let's go with that.

File: cosi/cosi.md [hunk 1]
```diff
@@ -258,64 +258,82 @@ steps below successfully.
...
++ the Announcement packet for the Announcement phase
++ the Commitment packet for the Commitment phase 
++ the Challenge packet for the Challenge phase 
++ the Response packet for the Response phase 
+
+The packet formats are described in details in the section XXX Packet Format XXX.
+
+For simplicity, we assume there is a designated leader who is responsible for
+collecting the shares and generating the signature. This leader could be any of
+the signers and is not trusted in any way. All participants are communicating
```
Comment (bford) [inline cosi/cosi.md]: We assume all participants...

File: cosi/cosi.md [hunk 2]
```diff
@@ -258,64 +258,82 @@ steps below successfully.
...
 round trips between the leader and the rest of the participants. 
 
 ### Announcement
 
-Upon the request to generate a signature on a statement S, the leader broadcasts
-an announcement message indicating the start of a signing process. It is up to
-the implementation to decide whether to send S itself during that phase or not. 
+Upon the request to generate a signature on a message M, the leader broadcasts
+an Announcement packet indicating the start of a signing process. This
+Announcement packet MUST contain the message M to sign.
```
Comment (bford) [inline cosi/cosi.md]: No, I think this is overly strong and not strictly necessary, and likely undesirable since it's likely to eliminate important pipelining opportunities.

I understand your concern that without knowing the message they're supposed to sign beforehand, cosigners that don't like the message will only be able to decide this in the response phase, and hence will require the protocol to restart (in the absence of an exception mechanism, which we haven't defined yet).  However, IF this proves to be a problem, then a better solution might be to add an exception mechanism rather than to add a strict requirement that the message be known in the announcement phase.

Also, I'm not convinced it's really likely to be a problem: in many/most scenarios I think we care about, the cosigners are typically supposed to be verifying a deterministic set of requirements that "should" be exactly the same as what the leader and every other cosigner are verifying themselves.  This is the case for blockchains (where everyone is supposed to agree on the validity rules), etc.  If an honest cosigner gets asked by a leader to sign something that doesn't check out, then that's probably an indication that the leader is seriously buggy or malicious, and it's perfectly reasonable for that cosigner to bail in a not-particularly-graceful way.  (And if the cosigner is dishonest and wants to DoS attack the signing process, we know the dishonest cosigner can do that regardless unless we jump into the added complexity of a restart-free scheme.)

In other words, for most of the applications I think we care about, cosigning should be a fully-automated process - it's not about some human sitting there navel-gazing and figuring out if they're comfortable cosigning or not, but about the cosigner simply double-checking that the thing the leader is asking to sign is indeed consistent with all the rules the leader itself should already have imposed on the message.  Being able to "gracefully decline" cosigning is not necessarily a priority, until/unless we actually identify some application for which that mode actually matters.

File: cosi/cosi.md [hunk 3]
```diff
@@ -258,64 +258,82 @@ steps below successfully.
...
 
 ### Commitment
 
-Upon the receipt of an announcement message or if the participant is the leader,
-each participant i generates a random secret r_i by hashing 32 bytes of
+Upon the receipt of an Announcement packet, each non-leader participant SHOULD
```
Comment (bford) [inline cosi/cosi.md]: Move this to the response phase; see above.

File: cosi/cosi.md (hunk 2 above)
Comment (bford) [reply cosi/cosi.md]: Another way to put it: any leader that's sending messages down for cosigning that the leader doesn't know that honest cosigners will sign is a broken leader.

File: cosi/cosi.md [hunk 4]
```diff
@@ -258,64 +258,82 @@ steps below successfully.
...
 
-The leader waits to receive the commitments R_i from the other participants for
-a certain time frame as defined by the application. After the timeout, the
-leader constructs the subset M of participants from whom he has received a
-commitment R_i and computes the sum R = SUM_{i in M}(R_i). The leader then
-computes SHA512(R || A || M) and interprets the resulting 64-byte digest as an
-integer c mod L.  The leader broadcasts c to all participants.
+The leader waits to receive the Commitment packets containing the R_i from the
+other participants for a certain time frame as defined by the application. After
+the timeout, the leader constructs the subset P' of participants from whom he
```
Comment (bford) [inline cosi/cosi.md]: After receiving all expected responses or the timeout...

File: cosi/cosi.md [hunk 5]
```diff
@@ -258,64 +258,82 @@ steps below successfully.
...
+has received a commitment R_i and computes the sum R = SUM_{i in P'}(R_i). The
+leader then computes SHA512(R || A || M) and interprets the resulting 64-byte
+digest as an integer c mod L.  The leader broadcasts c and R as a Challenge
```
Comment (bford) [inline cosi/cosi.md]: Why bother broadcasting c, when all cosigners can (and must anyway for security) recompute it anyway from R and the other inputs?

File: cosi/cosi.md [hunk 6]
```diff
@@ -258,64 +258,82 @@ steps below successfully.
...
-abort the protocol. The leader then computes the aggregate response s = SUM{i in
-M}(s_i) mod L and initializes a bitmask Z of size n to all zero. For each
-participant i who is present in N but not in M the leader sets the i-th bit of Z
-to 1, i.e., Z[i] = 1. The leader then forms the signature sig as the
-concatenation of the byte-encoded point R, the byte-encoded scalar s, and the
-bitmask Z. The resulting signature is of the form sig = R || s || Z and MUST be
-of length 32 + 32 + ceil(n/8) bytes.
+The leader waits to receive the Response packets containing the individual s_i
+from the other participants for a certain time frame as defined by the
+application. After the timeout, the leader checks if he received responses from
```
Comment (bford) [inline cosi/cosi.md]: After receiving all expected responses or the timeout...

Comment (bford) [reply cosi/cosi.md]: (shouldn't wait for the timeout if everyone responds, which hopefully should be the common case!)

File: cosi/cosi.md [hunk 7]
```diff
@@ -339,67 +357,79 @@ A leaf node is a node who has only one parent and no child nodes.
...
 
 // XXX What happens when a node does not receive any commitment from a child
 node. Does it contact the sub-nodes? 
 
 ### Challenge
 
-The leader computes the challenge c = H( R' || A || S) and BROADCASTS it
-down the tree. The leader also saves the bitmask Z' computed in the previous
-step. Upon reception, each leaf node executes the response step. 
+The leader computes the challenge c = H(R' || A || M) and BROADCASTS c and R' in
```
Comment (bford) [inline cosi/cosi.md]: Again, shouldn't be necessary to broadcast c, just R'.
